### ISSUE TITLE
Player falling damage inconsistent

### ISSUE THREAD

**Issue by mshaw** [minetest/minetest#3018]
Steps to reproduce
=====

1) Create new world name xx, seed xx, mapgen v7, game minimal
2) Start game with default mods
3) F10 issue command, "/teleport singleplayer 0 100 0"
4) Player appears in the air, falls and lands, any of the following occur:

Observed behaviour
=====

a) Player takes non-lethal damage (19 HP), or
b) player takes lethal damage and prompted to re-spawn, or
c) player takes lethal damage, console reads player has died, player doesn't die nor prompted to respawn, or
d) nothing.

Expected behaviour
=====

a) With this map seed the fall should be fatal,
b) the damage should be the same every time the player falls from this height, and
c) the player should need to re-spawn before continuing playing.

**Comment by PilzAdam** [comment]
Maybe the fall is interrupted by unloaded blocks.

**Comment by Wayward1** [comment]
I agree with PilzAdam; the only way I can reproduce said observations is if I hit unloaded blocks on the way down. Otherwise, the game behaves normally.

**Comment by paramat** [comment]
> Maybe the fall is interrupted by unloaded blocks.

^ This would affect fall speed.
@fmshaw can you reproduce this when all mapchunks are generated in the area of the fall?

**Comment by est31** [comment]
No its just probably that he hits unloaded blocks, that can be seen as bug.

**Comment by paramat** [comment]
That's not a bug, it's unavoidable. There was an issue about colliding with unloaded blocks. The player can't pass through because the block may be solid once it generates/loads.

**Comment by est31** [comment]
There are ways around it, e.g. the last velocity when the player falled can be saved when the actual velocity gets set to zero, and then gets added back onto the velocity when the block gets loaded.

There needs to be additional descisions to be made for example whether to allow the player to walk on the nodes (this will probably affect people mining at the bottom of the world as well), or what happens when the player jumps shortly before the block gets loaded again, or what happens when the player walks or jumps off the unloaded block, and continues their jump downwards or finds "secure land".

So yeah in total not really trivial.

**Comment by fmshaw** [comment]
I agree there are a lot of factors here... I'm guessing a "soft landing 
state" with deferred damage, paralysis and then inertia restored on 
block load is tricky to do and perhaps even not desirable.

I don't mind too much what the actual behavior is but feel consistency 
in result is important...

The player ends up standing on the same block.

The player takes the same amount of damage.

Regardless of loading of blocks.

**Comment by paramat** [comment]
I guess this was never much of an issue in mgv6 because of it's low terrain height. Teleporting to high up in an ungenerated area is what is causing mapblocks below to not be loaded. Also now we have more varied mapgens with floatlands and high mountains this will tend to happen more, especially when falling off floatlands.

Giving damage when player hits ignore can't be done because the fall may end up a fall into water, which is harmless.

**Comment by Wuzzy2** [comment]
Right, it is totally “expected behaviour” when a player falls down >100 nodes and has a good chance to not suffer any fall damage. :D

I don't care about unloaded chunks, dealing no fall damage for a >100 node drop is clearly a critical gameplay bug which should not be ignored.

One idea: Adding terminal velocity. At least it makes it less likely to hit an unloaded chunk but it may still be possible for slow servers.
Better idea: Freeze player when hitting an unloaded chunk (remembering the velocity) and unfreeze the player when the chunk is loaded, restoring fall velocity.
Even better idea: Combining  both ideas above.

Does anyone have better ideas at hand?

**Comment by Fixer-007** [comment]
Falling in minetest is buggy and needs rework (yes, it does). I can fall on flat ground while holding shift and don't take any damage. I can fall 1000 nodes down while looking up and survive.

**Comment by paramat** [comment]
Terminal velocity for a human skydiver is around 53m/s, it's highly likely a player will still hit unloaded blocks, especially if looking upwards, because blocks are sent according to the player's view.

**Comment by numberZero** [comment]
> because blocks are sent according to the player's view.

That’s what can be changed rather easily IMO. It should depend on player velocity if it is higher that normally.

> around 53m/s

MT need not to use exactly that value.

**Comment by paramat** [comment]
But if the blocks below are ungenerated their sending will be unavoidably delayed.
The approach of saving velocity when hitting an unloaded block is better.

**Comment by Fixer-007** [comment]
You should die no matter how you look, up or down or whatever, should be calculated properly, not just resetting damage each time.

**Comment by numberZero** [comment]
> The approach of saving velocity when hitting an unloaded block is better.

It is. But these approaches may be used together.

**Comment by paramat** [comment]
Yes we could have a terminal velocity, it would help slightly, it could be a new physics setting. It would be incorrect in space but that's a minor issue. Simple to add too.

**Comment by Wuzzy2** [comment]
Yes! Why not combining all these ideas?

First step is to decrease the likelihood the player can hit an unloaded chunk in the first place.
Second step is to make sure the velocity is not forgotten no matter what.

Specifically.

- Use terminal velocity
- Make the chunk loading also dependent on player's movement direction if the player is going fast. E.g. if the player is moving down at >10 blocks/second, loading chunks below the player should have a much higher priority over the blocks in looking direction
- In the (hopefully) unlikely event the player still hits an unloaded chunk, freeze the player and remember the velocity, until it is loaded

Yes, terminal velocity would be wrong in space, but I am arguing for this for technical reasons. 50 blocks/second would be very, very fast anyway (if Minetest could manage this speed).

**Comment by juhdanad** [comment]
Instead of working with velocities, you can also measure the distance of the fall. MC also does that. Each player/entity would have a fall_distance property, which is reset when the player swims, climbs or is on the ground. Landing on unloaded chunks would not reset this value.

**Comment by numberZero** [comment]
> which is reset when the player swims
> MC also does that

And 2 water nodes are enough in MC to avoid fall damage *completely*.
Freezing/unfreezing should work better. Implementation won’t be more complicated than proper (i.e. without such unwanted effects) implementation with fall distance.
